sources/csv/data.go

```diff
@@ -0,0 +1,284 @@
+	if len(tables) == 0 {
+		return fmt.Errorf("no tables found")
+	}
+	for i, table := range tables {
+		name := table.Table_name
+		if name == "" {
+			return fmt.Errorf("table number %d (0-indexed) does not have a name", i)
+		}
+		if len(table.File_patterns) == 0 {
+			return fmt.Errorf("no file path provided for table %s", name)
+		}
+		cols := table.Columns
+		if len(cols) == 0 {
+			return fmt.Errorf("`columns` field for table %s is empty", name)
+		}
+		for j, col := range cols {
+			if col.Column_name == "" || col.Type_name == "" {
+				return fmt.Errorf("please provide column_name and type_name in `columns` field at position %d (0-indexed)", j)
+			}
+		}
+	}
+	return nil
+}
+
+// SetRowStats calculates the number of rows per table.
+func SetRowStats(conv *internal.Conv, tables []Table) {
+	for _, table := range tables {
+		for _, filePath := range table.File_patterns {
+			count, err := getCSVRowCount(filePath)
+			if err != nil {
+				conv.Unexpected(fmt.Sprintf("Couldn't get number of rows for table %s", table.Table_name))
+				continue
+			}
+			conv.Stats.Rows[table.Table_name] += count
+		}
+	}
+}
+
+// getCSVRowCount returns the number of data rows in the CSV file.
+func getCSVRowCount(filePath string) (int64, error) {
+	count := int64(0)
+	csvFile, err := os.Open(filePath)
+	if err != nil {
+		fmt.Printf("can't read csv file: %s due to: %v\n", filePath, err)
+	}
+	r := csvReader.NewReader(csvFile)
+	for {
+		_, err := r.Read()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return 0, fmt.Errorf("can't read row from file: %s", filePath)
+		}
+		count++
+	}
+	// Exclude the first row, that is, column headers.
+	return count - 1, nil
+}
+
+// ProcessCSV writes data across the tables provided in the manifest file. Each table's data can be provided
+// across multiple CSV files hence, the manifest accepts a list of file paths in the input.
+func ProcessCSV(conv *internal.Conv, tables []Table) error {
+	for _, table := range tables {
+		// Populating just the table names in the conv for SrcSchema and SpSchema
+		// so the report for row stats is generated.
+		conv.SrcSchema[table.Table_name] = schema.Table{Name: table.Table_name}
+
+		// The map colDefs stores the mapping from column names to their final types.
+		colDefs := make(map[string]ddl.ColumnDef)
+		for _, col := range table.Columns {
+			ty, err := ToSpannerType(col.Type_name)
+			if err != nil {
+				return fmt.Errorf("can't map to spanner type: %v. Please use the data types as in your spanner database", err)
+			}
+			colDefs[col.Column_name] = ddl.ColumnDef{Name: col.Column_name, T: ty}
+		}
+		conv.SpSchema[table.Table_name] = ddl.CreateTable{Name: table.Table_name, ColDefs: colDefs}
+
+		for _, filePath := range table.File_patterns {
+			csvFile, err := os.Open(filePath)
+			if err != nil {
+				return fmt.Errorf(fmt.Sprintf("can't read csv file: %s due to: %v\n", filePath, err))
+			}
+			r := csvReader.NewReader(csvFile)
```

Comment:
What delimiters does this reader expect ? Is it only commas , or separated by double/single quotes. Can the delimiter be provided ?

Reply:
Currently only comma is there. But adding custom ones should be straight forward. I'll add it